<?xml version="1.0"?>
<robot name="car" xmlns:xacro="http://ros.org/wiki/xacro">

  <xacro:property name="pi" value="3.1415926535897931" />
  <xacro:property name="car_length" value="0.3762" />
  <xacro:property name="car_width" value="0.2124" />
  <xacro:property name="car_height" value="0.0762" />
  <xacro:property name="car_mass" value="0.7" />
  <xacro:property name="wheel_radius" value="0.0508" />
  <xacro:property name="wheel_length" value="0.0508" />
  <xacro:property name="wheel_mass" value="0.1" />
  <xacro:property name="capota_car_length" value="0.2746" />      <!-- x axis  -->
  <xacro:property name="capota_car_width" value="0.2124" />       <!-- y axis  -->
  <xacro:property name="capota_car_height" value="0.0508" />      <!-- z axis  -->
  <xacro:property name="capota_car_mass" value="0.3" />

  <xacro:property name="steering_cylinder_radius" value="0.01" /> 
  <xacro:property name="steering_cylinder_length" value="0.2124" /> 
  <xacro:property name="steering_cylinder_mass" value="0.05" />  

<!-- control parameters -->
  <xacro:property name="back_wheels_velocity_limit" value="100.0" />  
  <xacro:property name="steering_velocity_limit" value="5.0" />  


  <material name="green">
    <color rgba="0 1 0 1" />
  </material>
  <material name="light_green">
    <color rgba="0.1 1 0.2 1" />
  </material>
  <material name="yellow">
    <color rgba="1 1 0 1" />
  </material>
  <material name="grey">
    <color rgba="0.5 0.5 0.5 1"/>
  </material>

  <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
      <robotNamespace>/car</robotNamespace>
      <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
    </plugin>
  </gazebo>

<!-- For bringing colors to the main simulation -->
 <gazebo>
    <plugin name="gazebo_ros_material_plugin" filename="libgazebo_ros_api_plugin.so">
      <robotNamespace>/car</robotNamespace>
    </plugin>
  </gazebo>


  
  <!-- Dummy and Base Link -->
  <link name="dummy"> </link>
  <joint name="dummy_joint" type="fixed">
    <parent link="dummy" />
    <child link="base_link" />
  </joint>
  <link name="base_link">
    <visual>
      <geometry>
        <box size="${car_length} ${car_width} ${car_height}" />
      </geometry>
      <material name="green" />
    </visual>
    <collision>
      <geometry>
        <box size="${car_length} ${car_width} ${car_height}" />
      </geometry>
    </collision>
    <inertial>
      <mass value="${car_mass}" />
      <inertia
        ixx="${(car_mass/12) * (car_width**2 + car_height**2)}"
        ixy="0"
        ixz="0"
        iyy="${(car_mass/12) * (car_length**2 + car_height**2)}"
        iyz="0"
        izz="${(car_mass/12) * (car_length**2 + car_width**2)}" />
    </inertial>
  </link>

<!-- Capota of car -->
  <link name="capota">
    <visual>
      <geometry>
        <box size="${capota_car_length} ${capota_car_width} ${capota_car_height}" />
      </geometry>
      <material name="light_green" />
    </visual>
    <collision>
      <geometry>
        <box size="${capota_car_length} ${capota_car_width} ${capota_car_height}" />
      </geometry>
    </collision>
    <inertial>
      <mass value="${capota_car_mass}" />
      <inertia
        ixx="${(capota_car_mass/12) * (capota_car_width**2 + capota_car_height**2)}"
        ixy="0"
        ixz="0"
        iyy="${(capota_car_mass/12) * (capota_car_length**2 + capota_car_height**2)}"
        iyz="0"
        izz="${(capota_car_mass/12) * (capota_car_length**2 + capota_car_width**2)}" />
    </inertial>
  </link>
  <joint name="capota_joint" type="fixed">
    <parent link="base_link" />
    <child link="capota" />
    <origin xyz="${capota_car_length/2-car_length/2} 0 ${car_height/2+capota_car_height/2}"
      rpy="0 0 0" />
  </joint>


<!-- Steering cylinder axis -->
<link name="steering_cylinder_link">
    <visual>
      <geometry>
        <cylinder radius="${steering_cylinder_radius}" length="${steering_cylinder_length}"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <geometry>
        <cylinder radius="${steering_cylinder_radius}" length="${steering_cylinder_length}"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="${steering_cylinder_mass}"/>
      <inertia
        ixx="${(steering_cylinder_mass/12) * (3*steering_cylinder_radius**2 + steering_cylinder_length**2)}"
        ixy="0" ixz="0"
        iyy="${(steering_cylinder_mass/12) * (3*steering_cylinder_radius**2 + steering_cylinder_length**2)}"
        iyz="0"
        izz="${(steering_cylinder_mass/2) * steering_cylinder_radius**2}"/>
    </inertial>
  </link>

  <joint name="steering_cylinder_joint" type="revolute">
    <parent link="base_link"/>
    <child link="steering_cylinder_link"/>
    <origin xyz="${car_length/2-car_length/16} 0  ${-car_height/2}" rpy="${pi/2} 0 0"/>
    <axis xyz="0 1 0"/>
    <limit lower="${-pi/16}" upper="${pi/16}" effort="10.0" velocity="${steering_velocity_limit}"/>
  </joint>

 
  <!-- Wheels -->

  <!-- Front Left Wheel -->
  <link name="front_left_wheel_link">
    <visual>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
      </geometry>
      <material name="yellow"/>
    </visual>
    <collision>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="${wheel_mass}"/>
      <inertia
        ixx="${(wheel_mass/12) * (3*wheel_radius**2 + wheel_length**2)}"
        ixy="0" ixz="0"
        iyy="${(wheel_mass/12) * (3*wheel_radius**2 + wheel_length**2)}"
        iyz="0"
        izz="${(wheel_mass/2) * wheel_radius**2}"/>
    </inertial>
  </link>
  <joint name="front_left_wheel_joint" type="continuous">
    <parent link="steering_cylinder_link"/>
    <child link="front_left_wheel_link"/>
    <origin xyz="0 0 ${steering_cylinder_length/2+wheel_length/2}" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>
 
<!-- Front Right Wheel -->
 
  <link name="front_right_wheel_link">
    <visual>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
      </geometry>
      <material name="yellow"/>
    </visual>
    <collision>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="${wheel_mass}"/>
      <inertia
        ixx="${(wheel_mass/12) * (3*wheel_radius**2 + wheel_length**2)}"
        ixy="0" ixz="0"
        iyy="${(wheel_mass/12) * (3*wheel_radius**2 + wheel_length**2)}"
        iyz="0"
        izz="${(wheel_mass/2) * wheel_radius**2}"/>
    </inertial>
  </link>

  <joint name="front_right_wheel_joint" type="continuous">
    <parent link="steering_cylinder_link"/>
    <child link="front_right_wheel_link"/>
    <origin xyz="0 0 ${-steering_cylinder_length/2-wheel_length/2} " rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>



  <!-- back_left_wheel -->
  <link name="back_left_wheel_link">
    <visual>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_length}" />
      </geometry>
      <material name="yellow" />
    </visual>
    <collision>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_length}" />
      </geometry>
    </collision>
    <inertial>
      <mass value="${wheel_mass}" />
      <inertia
        ixx="${(wheel_mass/12) * (3*wheel_radius**2 + wheel_length**2)}"
        ixy="0"
        ixz="0"
        iyy="${(wheel_mass/12) * (3*wheel_radius**2 + wheel_length**2)}"
        iyz="0"
        izz="${(wheel_mass/2) * wheel_radius**2}" />
    </inertial>
  </link>
  <joint name="back_left_wheel_joint" type="continuous">
    <parent link="base_link" />
    <child link="back_left_wheel_link" />
    <origin xyz="${-car_length/2} ${+car_width/2+wheel_length/2} ${-car_height/2}" rpy="${pi/2} 0 0" />
    <axis xyz="0 0 1" />
    <limit effort="10.0" velocity="${back_wheels_velocity_limit}" />
  </joint>

  <!-- back_right_wheel -->
  <link name="back_right_wheel_link">
    <visual>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_length}" />
      </geometry>
      <material name="yellow" />
    </visual>
    <collision>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_length}" />
      </geometry>
    </collision>
    <inertial>
      <mass value="${wheel_mass}" />
      <inertia
        ixx="${(wheel_mass/12) * (3*wheel_radius**2 + wheel_length**2)}"
        ixy="0"
        ixz="0"
        iyy="${(wheel_mass/12) * (3*wheel_radius**2 + wheel_length**2)}"
        iyz="0"
        izz="${(wheel_mass/2) * wheel_radius**2}" />
    </inertial>
  </link>
  <joint name="back_right_wheel_joint" type="continuous">
    <parent link="base_link" />
    <child link="back_right_wheel_link" />
    <origin xyz="${-car_length/2} ${-car_width/2-wheel_length/2} ${-car_height/2}"
      rpy="${pi/2} 0 0" />
    <axis xyz="0 0 1" />
    <limit effort="10.0" velocity="${back_wheels_velocity_limit}" />
  </joint>

<!-- Camera -->
<link name="camera_link">
    <visual>
      <geometry>
        <box size="0.02 0.02 0.02"/> </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <geometry>
        <box size="0.02 0.02 0.02"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.01"/> <inertia ixx="0.00000833" ixy="0" ixz="0"
               iyy="0.00000833" iyz="0"
               izz="0.00000833"/>
    </inertial>
  </link>

  <joint name="camera_joint" type="fixed">
    <parent link="capota"/>
    <child link="camera_link"/>
    <origin xyz="${capota_car_width/2} 0 ${capota_car_height/2 + 0.002}" rpy="0 0 0"/>
  </joint>

<gazebo reference="camera_link">
    <sensor type="camera" name="car_camera_sensor">
      <update_rate>30.0</update_rate> <camera name="car_camera">
        <horizontal_fov>1.39626</horizontal_fov> <image>
          <width>640</width>
          <height>480</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.05</near>
          <far>10.0</far>
        </clip>
        <noise>
          <type>gaussian</type>
          <mean>0.0</mean>
          <stddev>0.007</stddev>
        </noise>
      </camera>
      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <robotNamespace>/car</robotNamespace> <alwaysOn>true</alwaysOn>
        <updateRate>30.0</updateRate>
        <cameraName>car_camera</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
        <frameName>camera_link_optical</frameName> <hackBaseline>0.07</hackBaseline> <distortionK1>0.0</distortionK1>
        <distortionK2>0.0</distortionK2>
        <distortionK3>0.0</distortionK3>
        <distortionT1>0.0</distortionT1>
        <distortionT2>0.0</distortionT2>
      </plugin>
    </sensor>
  </gazebo>
  <transmission name="back_left_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="back_left_wheel_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="back_left_wheel_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="back_right_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="back_right_wheel_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="back_right_wheel_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="front_steer_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="steering_cylinder_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="front_steer_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>


  
  <!-- Gazebo friction parameters -->
  <gazebo reference="base_link">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>
  <gazebo reference="back_right_wheel_link">
    <mu1>25.0</mu1>
    <mu2>25.0</mu2>
  </gazebo>
  <gazebo reference="back_left_wheel_link">
    <mu1>25.0</mu1>
    <mu2>25.0</mu2>
  </gazebo>
  <gazebo reference="front_right_wheel_link">
    <mu1>25.0</mu1>
    <mu2>25.0</mu2>
  </gazebo>
  <gazebo reference="front_left_wheel_link">
    <mu1>25.0</mu1>
    <mu2>25.0</mu2>
  </gazebo>
</robot>